sim/battle-queue.ts

```diff
@@ -0,0 +1,366 @@
+		for (const [i, curAction] of this.entries()) {
+			if (curAction === action) {
+				this.splice(i, 1);
+				break;
+			}
+		}
+		action.sourceEffect = sourceEffect;
+		action.order = 3;
+		this.unshift(action);
+	}
+
+	/**
+	 * Changes a pokemon's action, and inserts its new action
+	 * in priority order.
+	 *
+	 * You'd normally want the OverrideAction event (which doesn't
+	 * change priority order).
+	 */
+	changeAction(pokemon: Pokemon, action: ActionChoice) {
+		this.cancelAction(pokemon);
+		if (!action.pokemon) action.pokemon = pokemon;
+		this.insertChoice(action);
+	}
+
+	addChoice(choices: ActionChoice | ActionChoice[]) {
+		if (!Array.isArray(choices)) choices = [choices];
+		for (const choice of choices) {
+			this.push(...this.resolveAction(choice));
+		}
+	}
+
+	willAct() {
+		for (const action of this) {
+			if (['move', 'switch', 'instaswitch', 'shift'].includes(action.choice)) {
+				return action;
+			}
+		}
+		return null;
+	}
+
+	willMove(pokemon: Pokemon) {
+		if (pokemon.fainted) return false;
+		for (const action of this) {
+			if (action.choice === 'move' && action.pokemon === pokemon) {
+				return action;
+			}
+		}
+		return null;
+	}
+
+	cancelAction(pokemon: Pokemon) {
+		const oldLength = this.length;
+		for (let i = 0; i < this.length; i++) {
+			if (this[i].pokemon === pokemon) {
+				this.splice(i, 1);
+				i--;
+			}
+		}
+		return this.length !== oldLength;
+	}
+
+	cancelMove(pokemon: Pokemon) {
+		for (const [i, action] of this.entries()) {
+			if (action.choice === 'move' && action.pokemon === pokemon) {
+				this.splice(i, 1);
+				return true;
+			}
+		}
+		return false;
+	}
+
+	willSwitch(pokemon: Pokemon) {
+		for (const action of this) {
+			if (['switch', 'instaswitch'].includes(action.choice) && action.pokemon === pokemon) {
+				return action;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Inserts the passed action into the action queue when it normally
+	 * would have happened (sorting by priority/speed), without
+	 * re-sorting the existing actions.
+	 */
+	insertChoice(choices: ActionChoice | ActionChoice[], midTurn: boolean = false) {
+		if (Array.isArray(choices)) {
+			for (const choice of choices) {
+				this.insertChoice(choice);
+			}
+			return;
+		}
+		const choice = choices;
+
+		if (choice.pokemon) {
+			choice.pokemon.updateSpeed();
+		}
+		const actions = this.resolveAction(choice, midTurn);
+		for (const [i, curAction] of this.entries()) {
+			if (BattleQueue.comparePriority(actions[0], curAction) < 0) {
+				this.splice(i, 0, ...actions);
+				return;
+			}
+		}
+		this.push(...actions);
+	}
+
+	clear() {
+		this.splice(0);
+	}
+
+	debug() {
+		return this.map(action =>
+			// @ts-ignore
+			`${action.order || ''}:${action.priority || ''}:${action.speed || ''}:${action.subOrder || ''} - ${action.choice}${action.pokemon ? ' ' + action.pokemon : ''}${action.move ? ' ' + action.move : ''}`
+		).join('\n') + '\n';
+	}
+
+	sort(DO_NOT_USE?: unknown) {
```

Comment:
Whats this argument for? I'm not aware of `battle.sortQueue` previously taking an argument.

Reply:
It's for TypeScript. Without it, TypeScript complains about `sort` not matching the signature of `Array#sort`

Reply:
It's nice to be able to throw an error if someone tries to pass a comparator, though.

Reply:
I think if you type it as `DO_NOT_USE?: never` then TS will complain if you pass it anything.

Reply:
I'd imagine that if I did that, TypeScript would complain (the same way it complained about having the signature be `sort()`), but apparently it's not complaining about that anymore. I wonder why the inconsistency, but I'm not complaining.